### ISSUE TITLE
Session gets lost if nesting promises

### ISSUE THREAD

**Issue by alexandruluca** [othiym23/node-continuation-local-storage#116]
I'm using this package to create a session for the current user so I can access session inside nested promises, including mongoose plugins and so on. For whatever reason, the namespace exists too quick and the session get's lost.

I debugged your code a little bit and I came up with this solution for my problem. In context.js line 44 in the Namespace.prototype.run method I patched it like this

```javascript
Namespace.prototype.run = function (fn) {
  var context = this.createContext();
  this.enter(context);
  try {
    var contextFnResult = fn(context);
	  return context;
  }
  catch (exception) {
    if (exception) {
      exception[ERROR_SYMBOL] = context;
    }
    this.exit(context);
    throw exception;
  }
  finally {
    context.res.on('finish', function() {
		this.exit(context);
    });
  }
};
```

Basically I simply exit the namespace on an error and always exit on response finish

```javascript
context.res.on('finish', function() {
		this.exit(context);
    });
```
This way I am sure that my session is active untill I send back my response to the client

This works for me as intended. Are there any cons to this approach? Looking forward to your feedback

**Comment by AleksMeshkov** [comment]
Had to use almost the same approach. 

https://github.com/wjamro/node-continuation-local-storage/commit/3bbb6de915176eb5397c957b99e9a69ec457a182

**Comment by zsteinkamp** [comment]
@AleksMeshkov have you opened a pull request? This will be a problem for anyone using CLS with express.

**Comment by blacksonic** [comment]
@othiym23 Do we need a test for @AleksMeshkov s fix in a pull request?

**Comment by Qard** [comment]
Have you tried using `ns.bindEmitter(res)`?

Seems likely there is a branch of the execution tree that CLS is not aware of, resulting in the tree resolving early and exiting when some activity is still pending. The most likely cause of that is event emitters as they are not patched automatically.
